clang/lib/Sema/Metafunctions.cpp

```diff
@@ -1252,6 +1257,66 @@ bool map_decl_to_entity(APValue &Result, Sema &S, EvalFn Evaluator,
   llvm_unreachable("unknown reflection kind");
 }
 
+template <typename DT>
+bool getQualifiedName(APValue &Result, Sema &S, EvalFn Evaluator, DT& declType)
+{
+    SmallString<128> QualifiedNameBuffer;
+    llvm::raw_svector_ostream OS(QualifiedNameBuffer);
+    declType.printQualifiedName(OS);
+    StringRef Name = QualifiedNameBuffer.str();
+    if( makeCString(Result, Name, S.Context, Evaluator) )
+    {
+      llvm_unreachable("failed to create qualified name string");
+      return true;
+    }
+    return false;
+}
+
+bool qualified_name_of(APValue &Result, Sema &S, EvalFn Evaluator, QualType ResultTy,
+             SourceRange Range, ArrayRef<Expr *> Args) {
+  assert(Args[0]->getType()->isReflectionType());
+  assert(ResultTy ==
+         S.Context.getPointerType(S.Context.getConstType(S.Context.CharTy)));
+
+  APValue R;
+  if (!Evaluator(R, Args[0], true))
+    return true;
+
+  switch (R.getReflection().getKind()) {
+  case ReflectionValue::RK_type: { 
+    QualType QT = R.getReflectedType(); 
+    return SetAndSucceed(Result, getTypeName(S.Context, Evaluator, QT,
+                                             /*emptyIfUnnamed=*/true));
+  }
+  case ReflectionValue::RK_declaration: { 
+    const ValueDecl *VD = cast<ValueDecl>(R.getReflectedDecl());
+    return getQualifiedName(Result, S, Evaluator, *VD);
+  }
+  case ReflectionValue::RK_template: {  
+    const TemplateDecl *TD = cast<TemplateDecl>(R.getReflectedTemplate().getAsTemplateDecl());
+    return getQualifiedName(Result, S, Evaluator, *TD);
+  }
+  case ReflectionValue::RK_const_value: {
+    if (makeCString(Result, "", S.Context, Evaluator))
+      llvm_unreachable("failed to create empty string");
+    return false;
+  }
+  case ReflectionValue::RK_namespace: {
+    const NamespaceDecl *ND = cast<NamespaceDecl>(R.getReflectedNamespace());
```

Comment:
This, surprisingly, actually isn't quite right - in addition to a `NamespaceDecl`, `getReflectedNamespace()` might be a `NamespaceAliasDecl *` or a `TranslationUnitDecl *` (i.e., for reflections like `^::` of the global scope). Neither of these has `NamespaceDecl` as an ancestor base class, so this `cast<NamespaceDecl>` will crash with an assert-failure.

I think what you'll want to do here is:
1. Get `R.getReflectedNamespaced()` as a `Decl *`.
2. Check the special case of whether it's a `TranslationUnitDecl`. If it is, return the empty string (the global namespace isn't a "declared entity", so P2996 says this should return empty).
3. Otherwise cast it to a `NamedDecl *` (which is a base class of both `NamespaceDecl` and `NamespaceAliasDecl`), and pass that to `getQualifiedName`.

We should probably add test cases for `^::` and for namespace aliases as well.